website/blog/2022-08-22-narrative-modeling.md

```diff
@@ -0,0 +1,236 @@
+
+- Navigating an analytical database should be intuitive and educational
+- Traversing the underlying code for an analytical database should be intuitive and stakeholder-friendly
+- Adding to an analytical database should be intuitive and scaleable
+
+Out of those principles came three primary conventions: event tables, details tables, and entity schemas.
+
+## Event tables
+
+First, the backbone of a good narrative is the plot: what happened? In our example, *How many orders came in last week?*
+
+To answer this question, we still need one row per order placed, with timestamps so that we can filter down to specific time periods (e.g. last week). But what should we name it?
+
+Every principle of narrative modeling included the work “intuitive”. With that in mind, let’s call this table `order_placed`.
+
+In narrative modeling, we’d call `order_placed`  an *event table*. Event tables have the naming convention of `subject_verb` . 
+
+If this looks similar to `fct_` tables, it is! However, the benefit of `order_placed`over `fct_orders` appears when you need an additional order-level event. For instance, *How many orders were fulfilled last week?*
+
+- In dimensional modeling, you already used `fct_orders`, so you likely would need to rename that model. This can result in naming conventions like `fct_order_placement` and `fct_order_fulfillment` which can feel like reporting on two separate entities rather than two things that can happen to orders.
+- In narrative modeling, the sky’s the limit as we stand up `order_fulfilled` alongside `order_placed` with room for `order_cancelled` and all matter of other steps along an order’s flow.
+
+Now we can model out the narrative thrust of an order, but events can be a little dry without a fleshed-out protagonist. How would we describe our hero, a single order?
+
+## Details tables
+
+In narrative modeling, details tables are where a user can expect to find more descriptive information about a particular business entity. 
+
+In the example of orders, an `order_details` table might contain fields like:
+
+- `customer_id`
+- `items_ordered`
+- `payment_method`
+
+Initially, it might seem as though these should just be fields on `order_placed` and, for those examples, it’s a reasonable point[^1]. Consider instead the following:
+
+- `items_fulfilled`
+- `days_to_fulfillment`
+
+One can imagine a stakeholder wanting to know all five of these data points, but the context for them comes from multiple events in the order’s flow. A`_details` table allows us to assemble a broad bank of knowledge about a given entity (in this case, an order) in one place. If data refresh is performant enough, this could even be the place to surface true snapshot fields like `status` or`current_location`.
```

Comment:
```suggestion
One can imagine a stakeholder wanting to know all five of these data points, but the context for them comes from multiple events in the order’s flow. A `_details` table allows us to assemble a broad bank of knowledge about a given entity (in this case, an order) in one place. If data refresh is performant enough, this could even be the place to surface true snapshot fields like `status` or`current_location`.
```